File: LeetCodeTasks/3003_maximize-the-number-of-partitions-after-operations/solutions/python3/solution.py
class Solution:
    def maxPartitionsAfterOperations(self, s: str, k: int) -> int:
        from collections import defaultdict, deque
        
        # Function to calculate the number of partitions without any modification
        def count_partitions(s):
            n = len(s)
            i = 0
            parts = 0
            while i < n:
                char_count = defaultdict(int)
                distinct = 0
                j = i
                while j < n and (distinct < k or (distinct == k and s[j] in char_count)):
                    if char_count[s[j]] == 0:
                        distinct += 1
                    char_count[s[j]] += 1
                    if distinct > k:
                        break
                    j += 1
                parts += 1
                i = j
            return parts
        
        # Early exit: if k is large enough to take the whole string in one go
        if k >= 26:
            return 1
        
        # Calculate partitions without any changes
        max_partitions = count_partitions(s)
        
        # Optimally choosing to change one character to maximize partitions
        n = len(s)
        if n == 1:
            return 1
        
        # We'll try changing each character to any other character
        # and see if it improves the number of partitions
        for i in range(n):
            original_char = s[i]
            for c in range(26):
                new_char = chr(ord('a') + c)
                if new_char == original_char:
                    continue
                new_s = s[:i] + new_char + s[i+1:]
                max_partitions = max(max_partitions, count_partitions(new_s))
        
        return max_partitions

File: LeetCodeTasks/3003_maximize-the-number-of-partitions-after-operations/solutions/c/solution.c
#include <string.h>
#include <stdlib.h>
#include <limits.h>

int maxPartitionsAfterOperations(char* s, int k) {
    int n = strlen(s);
    int max_partitions = 1;
    char original_char;
    
    // Helper function to count partitions with at most k distinct characters
    int count_partitions(char* str) {
        int partitions = 0;
        int l = 0, distinct = 0, freq[26] = {0};
        
        for (int r = 0; r < n; r++) {
            if (freq[str[r] - 'a']++ == 0) distinct++;
            while (distinct > k) {
                if (--freq[str[l] - 'a'] == 0) distinct--;
                l++;
            }
            partitions++;
        }
        
        return partitions;
    }
    
    // First, evaluate the partition count without any changes
    max_partitions = count_partitions(s);
    
    // Check changing each character if it potentially improves the partition count
    for (int i = 0; i < n; i++) {
        original_char = s[i];
        for (char c = 'a'; c <= 'z'; c++) {
            if (c != original_char) {
                s[i] = c;
                int partitions = count_partitions(s);
                if (partitions > max_partitions) {
                    max_partitions = partitions;
                }
            }
        }
        s[i] = original_char; // Restore original character
    }
    
    return max_partitions;
}
